In [8]:
# 这是初始化链接信息，请忽略
%load_ext sql
%config SqlMagic.autocommit=False
%config SqlMagic.displaylimit=20
%sql $KYUUBI_URL

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


## **一、 基础查询与过滤**
### **1. 基础过滤与排序**
<font color="red">题目：</font> 查询所有状态为 paid（已支付）且金额大于 500 元的订单，按金额从高到低排序，只展示前 5 条。

<font color="#008000">考察点：</font>WHERE 多条件过滤 (AND)、ORDER BY 降序 (DESC)、LIMIT。

<font color="orange">答案：</font>

In [6]:
%%sql
SELECT
    order_id       AS order_id
  , user_id        AS user_id
  , total_amount   AS total_amount
  , payment_method AS payment_method
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
      status = 'paid'
  AND CAST(total_amount AS DOUBLE) > 500
ORDER BY
    CAST(total_amount AS DOUBLE) DESC
LIMIT 5;

ModuleNotFoundError: No module named 'pandas'

In [7]:
%%sql
SELECT
    order_id       AS order_id
  , user_id        AS user_id
  , total_amount   AS total_amount
  , payment_method AS payment_method
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
      status = 'paid'
  AND CAST(total_amount AS DOUBLE) > 500
ORDER BY
    CAST(total_amount AS DOUBLE) DESC
LIMIT 5;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,user_id,total_amount,payment_method
202501150002,10007,5000.00,credit
202501200001,10002,5000.00,credit
202501060002,10007,4000.00,credit
202501070002,10007,3500.00,credit
202501070001,10002,3000.00,credit


### **2. 字符串模糊匹配**
<font color="red">题目：</font> 找出所有商品名称中包含“手机”或者以“智能”开头的商品明细。

<font color="#008000">考察点：</font>LIKE 操作符，% (任意字符) 通配符的使用，OR 逻辑。

<font color="orange">答案：</font>

In [3]:
%%sql
SELECT
    product_id   AS product_id
  , product_name AS product_name
  , category     AS category
  , unit_price   AS unit_price
FROM
    dataspire_catalog.db_dev.fact_order_items
WHERE
     product_name LIKE '%手机%'
  OR product_name LIKE '智能%'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


product_id,product_name,category,unit_price
5001,智能手机Pro,手机数码,500.00
5009,智能电视,家用电器,5000.00
5012,智能手表,智能穿戴,350.00
5020,高端手机,手机数码,3000.00


###  **3. 正则表达式匹配**

<font color="red">题目：</font> 筛选出订单号中以 "202501" 开头且第三四位是 "0[1-5]" (上旬) 的订单。

<font color="#008000">考察点：</font>RLIKE 正则表达式匹配，^ (开头), [] (字符集合) 的使用。

<font color="orange">答案：</font>

In [4]:
%%sql
SELECT
    order_id   AS order_id
  , order_date AS order_date
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_id RLIKE '^2025010[1-5].*'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,order_date
202501010001,2025-01-01
202501020001,2025-01-02
202501030001,2025-01-03
202501050001,2025-01-05
202501020002,2025-01-02
202501030002,2025-01-03
202501040001,2025-01-04
202501050002,2025-01-05


### **4. 抽样查询**

<font color="red">题目：</font> 随机抽取 10% 的订单数据进行快速预览。

<font color="#008000">考察点：</font>TABLESAMPLE 子句，大数据量下的快速采样技巧。

<font color="orange">答案：</font>

In [5]:
%%sql
SELECT
    order_id     AS order_id
  , total_amount AS total_amount
FROM
    dataspire_catalog.db_dev.fact_orders TABLESAMPLE (10 PERCENT)
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,total_amount


## **二、 聚合函数与分组**
<font color="red">题目：</font> 统计每个下单渠道 (channel) 的总订单金额，并仅显示总金额高于 5000 的渠道。注意金额字段为字符串类型。

<font color="#008000">考察点：</font>GROUP BY 分组，SUM 聚合，CAST 类型转换，HAVING 子句过滤分组结果。

<font color="orange">答案：</font>

In [6]:
%%sql
SELECT
    t.channel                           AS channel
  , SUM(CAST(t.total_amount AS DOUBLE)) AS total_gmv
FROM
    dataspire_catalog.db_dev.fact_orders t
WHERE
    status = 'paid'
GROUP BY
    channel
HAVING
    SUM(CAST(total_amount AS DOUBLE)) > 5000
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


channel,total_gmv
web,10080.0
app,31270.0


## **三、 多表连接 (JOIN)**
### **1. INNER JOIN (内连接)**

<font color="red">题目：</font> 查询用户姓名 (user_name) 及其购买的商品名称 (product_name)。表结构关联路径为：dim_users -> fact_orders -> fact_order_items。

<font color="#008000">考察点：</font>取交集，最常用的连接方式。

<font color="orange">答案：</font>

In [7]:
%%sql
SELECT
    u.user_name    AS user_name
  , i.product_name AS product_name
FROM
    dataspire_catalog.db_dev.dim_users u
        INNER JOIN dataspire_catalog.db_dev.fact_orders o
                   ON u.user_id = o.user_id
        INNER JOIN dataspire_catalog.db_dev.fact_order_items i
                   ON o.order_id = i.order_id
WHERE
    o.status = 'paid'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_name,product_name
王五,运动鞋
Lisa,化妆品
李四,办公桌椅
Mike,高端相机
Emma,咖啡机
Emma,扫地机器人
赵六,蓝牙耳机
Mike,高端手机
Mike,高端电脑
张三,平板电脑


### **2. LEFT OUTER JOIN (左外连接)**

<font color="red">题目：</font> 列出所有用户，即使他们没有下过单（显示0元）。

<font color="#008000">考察点：</font>以左表为主，右表补全。常用于“主表 + 属性表”场景。

<font color="orange">答案：</font>

In [8]:
%%sql
SELECT
    u.user_id                                        AS user_id
  , u.user_name                                      AS user_name
  , COALESCE(SUM(CAST(o.total_amount AS DOUBLE)), 0) AS total_spent
FROM
    dataspire_catalog.db_dev.dim_users u
        LEFT JOIN dataspire_catalog.db_dev.fact_orders o
                  ON u.user_id = o.user_id
                      AND o.status = 'paid'
GROUP BY
    u.user_id, u.user_name
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_name,total_spent
10001,张三,3000.0
10002,李四,11500.0
10007,Mike,18000.0
10008,Lisa,1350.0
10004,赵六,1600.0
10005,Emma,4500.0
10006,John,800.0
10003,王五,600.0


### **3. RIGHT OUTER JOIN (右外连接)**

<font color="red">题目：</font> 查询所有用户，并附上他们的订单信息。即使用户从未下过单，也要列出该用户（订单字段为 NULL）。

<font color="#008000">考察点：</font>以右表为主。逻辑同 Left Join，只是表的位置互换。

<font color="Blue">注：</font>在 Spark 中通常建议统一用 Left Join (交换表位置) 以避免优化器混淆，但语法必须掌握。

<font color="orange">答案：</font>

In [9]:
%%sql
SELECT
    u.user_id      AS user_id
  , u.user_name    AS user_name
  , o.order_id     AS order_id -- 若用户无订单，此处为 NULL
  , o.total_amount AS total_amount -- 若用户无订单，此处为 NULL
FROM
    dataspire_catalog.db_dev.fact_orders o
        RIGHT JOIN dataspire_catalog.db_dev.dim_users u
                   ON o.user_id = u.user_id
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_name,order_id,total_amount
10001,张三,202501010001,500.00
10001,张三,202501020001,300.00
10001,张三,202501030001,200.00
10001,张三,202501100001,800.00
10001,张三,202501150001,1200.00
10002,李四,202501050001,2000.00
10002,李四,202501060001,1500.00
10002,李四,202501070001,3000.00
10002,李四,202501200001,5000.00
10007,Mike,202501040001,3000.00


### **4. FULL OUTER JOIN (全外连接)**

<font color="red">题目：</font> 进行数据对账。找出所有的订单和所有的用户。匹配的：显示在一起。有订单无用户的：显示订单，用户列为 NULL。有用户无订单的：显示用户，订单列为 NULL。

<font color="#008000">考察点：</font>取并集。常用于发现数据不一致（孤儿数据）。

<font color="orange">答案：</font>

In [10]:
%%sql
SELECT
    COALESCE(o.user_id, u.user_id) AS common_user_id -- 处理两边的 ID
  , o.order_id                     AS order_id
  , u.user_name                    AS user_name
  , CASE
        WHEN o.order_id IS NULL THEN '用户无订单'
        WHEN u.user_id IS NULL THEN '订单无用户'
        ELSE '匹配成功'
        END                        AS match_status
FROM
    dataspire_catalog.db_dev.fact_orders o
        FULL JOIN dataspire_catalog.db_dev.dim_users u ON o.user_id = u.user_id
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


common_user_id,order_id,user_name,match_status
10001,202501010001,张三,匹配成功
10001,202501020001,张三,匹配成功
10001,202501030001,张三,匹配成功
10001,202501100001,张三,匹配成功
10001,202501150001,张三,匹配成功
10002,202501050001,李四,匹配成功
10002,202501060001,李四,匹配成功
10002,202501070001,李四,匹配成功
10002,202501200001,李四,匹配成功
10007,202501040001,Mike,匹配成功


### **5. SEMI JOIN (半连接)**

<font color="red">题目：</font> 查询下过单的用户列表（去重）。只需要用户信息，不需要订单详情。只要该用户在订单表出现过至少一次即可。

<font color="#008000">考察点：</font>IN 子查询的高效替代版。注意：Select 子句中只能出现左表的字段，不能选右表字段。性能通常优于 DISTINCT + JOIN。

<font color="orange">答案：</font>

In [11]:
%%sql
-- 逻辑解读：返回 dim_users 中那些能在 fact_orders (paid) 中找到匹配项的行。
-- 结果：每个用户只出现一次（即使下了100单），且不包含订单表的任何列。
SELECT
    u.user_id   AS user_id
  , u.user_name AS user_name
  , u.city      AS city
FROM
    dataspire_catalog.db_dev.dim_users u
    SEMI JOIN dataspire_catalog.db_dev.fact_orders o
ON u.user_id = o.user_id
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_name,city
10001,张三,北京
10002,李四,上海
10007,Mike,北京
10008,Lisa,广州
10004,赵六,广州
10005,Emma,深圳
10006,John,上海
10003,王五,北京


### **6. ANTI JOIN (左反连接)**

<font color="red">题目：</font> 查询从未下过单的用户（潜在流失用户或注册未转化用户）。

<font color="#008000">考察点：</font>NOT IN 或 NOT EXISTS 的高效替代版。同样只能选左表字段。

<font color="orange">答案：</font>

In [12]:
%%sql
-- 逻辑解读：返回 dim_users 中那些在 fact_orders 中**找不到**任何匹配项的行。
-- 结果：全是没下过单的用户。
SELECT
    u.user_id       AS user_id
  , u.user_name     AS user_name
  , u.register_date AS register_date
FROM
    dataspire_catalog.db_dev.dim_users u
    ANTI JOIN dataspire_catalog.db_dev.fact_orders o
ON u.user_id = o.user_id
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_name,register_date


### **7. CROSS JOIN (交叉连接/笛卡尔积)**

<font color="red">题目：</font> 生成测试数据，或者做“维度的全组合”。例如：列出每个城市与每种支付方式的所有可能组合（不管实际上有没有发生）。

<font color="#008000">考察点：</font>行数 = 左表行数 × 右表行数。警告：生产环境慎用，极易导致数据爆炸！

<font color="orange">答案：</font>

In [13]:
%%sql
-- 如果有 5 个城市，3 种支付方式，结果将有 15 行 (5*3)。
-- 每一行代表一种理论上的“城市 - 支付方式”组合。
SELECT
    c.city           AS city
  , p.payment_method AS payment_method
  , 'All_Combo'      AS scenario
FROM
    (
    SELECT DISTINCT
        city AS city
    FROM
        dataspire_catalog.db_dev.dim_users
    ) c
        CROSS JOIN (
                   SELECT DISTINCT
                       payment_method AS payment_method
                   FROM
                       dataspire_catalog.db_dev.fact_orders
                   ) p
LIMIT 10
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


city,payment_method,scenario
深圳,wechat,All_Combo
深圳,credit,All_Combo
深圳,alipay,All_Combo
上海,wechat,All_Combo
上海,credit,All_Combo
上海,alipay,All_Combo
北京,wechat,All_Combo
北京,credit,All_Combo
北京,alipay,All_Combo
广州,wechat,All_Combo


## **四、 窗口函数**
### **1. 排序函数**

<font color="red">题目：</font> 在每个 city 内部，按照 register_date 对用户进行排序，生成一个从 1 开始的连续序号。如果注册日期相同，按 user_id 排序

<font color="#008000">考察点：</font>ROW_NUMBER()，RANK()，DENSE_RANK()，PARTITION BY 分组，ORDER BY 多列排序。

<font color="orange">答案：</font>

In [14]:
%%sql
SELECT
    user_id                                                               AS user_id
  , city                                                                  AS city
  , register_date                                                         AS register_date
  , ROW_NUMBER() OVER (PARTITION BY city ORDER BY register_date, user_id) AS city_row_number
  , RANK() OVER (PARTITION BY city ORDER BY register_date, user_id)       AS city_rank
  , DENSE_RANK() OVER (PARTITION BY city ORDER BY register_date, user_id) AS dense_rank
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,city,register_date,city_row_number,city_rank,dense_rank
10005,深圳,2024-08-22,1,1,1
10002,上海,2024-03-20,1,1,1
10006,上海,2024-11-30,2,2,2
10001,北京,2024-01-15,1,1,1
10007,北京,2024-05-18,2,2,2
10003,北京,2024-06-10,3,3,3
10004,广州,2025-01-05,1,1,1
10008,广州,2025-02-01,2,2,2


### **2. 直方图/分桶统计**

<font color="red">题目：</font> 将全体用户按照 total_spent 从高到低均匀分成 4 个梯队（四分位），标记每个用户属于哪个梯队（1-4）。

<font color="#008000">考察点：</font>NTILE(n)，用于二八分析或分层运营。

<font color="orange">答案：</font>

In [15]:
%%sql
SELECT
    user_id                                   AS user_id
  , total_spent                               AS total_spent
  , NTILE(4) OVER (ORDER BY total_spent DESC) AS spending_quartile
FROM
    (
    SELECT
        o.user_id           AS user_id
      , SUM(o.total_amount) AS total_spent
    FROM
        dataspire_catalog.db_dev.fact_orders o
            INNER JOIN dataspire_catalog.db_dev.dim_users u
                       ON o.user_id = o.user_id
    GROUP BY
        1
    ) t
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,total_spent,spending_quartile
10007,144000.0,1
10002,92000.0,1
10005,36000.0,2
10001,24000.0,2
10004,12800.0,3
10008,10800.0,3
10003,8000.0,4
10006,6400.0,4


### **3. 相对位置百分比 (PERCENT_RANK)**

<font color="red">题目：</font> 计算每个订单金额在所有订单中的相对排名百分比（0.0 到 1.0）。

<font color="#008000">考察点：</font>PERCENT_RANK()，评估数据在整体中的相对位置。

<font color="orange">答案：</font>

In [16]:
%%sql
SELECT
    order_id                                    AS order_id
  , total_amount                                AS total_amount
  , PERCENT_RANK() OVER (ORDER BY total_amount) AS amount_percentile
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,total_amount,amount_percentile
202501080001,1000.00,0.0
202501150001,1200.00,0.04
202501060001,1500.00,0.08
202501180001,1500.00,0.08
202501030001,200.00,0.16
202501050001,2000.00,0.2
202501280001,2000.00,0.2
202501050002,2500.00,0.28
202502030001,280.00,0.32
202501020001,300.00,0.36


### **4. 累积分布 (CUME_DIST)**

<font color="red">题目：</font> 计算每个订单金额的累积分布值（小于等于当前金额的订单比例）。

<font color="#008000">考察点：</font>CUME_DIST()，与 PERCENT_RANK 类似但计算公式不同（包含当前值）。

<font color="orange">答案：</font>

In [17]:
%%sql
SELECT
    order_id                                 AS order_id
  , total_amount                             AS total_amount
  , CUME_DIST() OVER (ORDER BY total_amount) AS cum_dist
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,total_amount,cum_dist
202501080001,1000.00,0.038461538461538464
202501150001,1200.00,0.07692307692307693
202501060001,1500.00,0.15384615384615385
202501180001,1500.00,0.15384615384615385
202501030001,200.00,0.19230769230769232
202501280001,2000.00,0.2692307692307692
202501050001,2000.00,0.2692307692307692
202501050002,2500.00,0.3076923076923077
202502030001,280.00,0.34615384615384615
202501020001,300.00,0.38461538461538464


### **5. 获取前后一行数据 (LAG)**

<font color="red">题目：</font> 计算每个用户本次订单金额与上一笔订单金额的差值。如果是第一单，差值记为 0。

<font color="#008000">考察点：</font>LAG(col, offset, default)，计算环比变化。

<font color="orange">答案：</font>

In [18]:
%%sql
SELECT
    user_id                                                                                  AS user_id
  , order_date                                                                               AS order_date
  , total_amount                                                                             AS total_amount
  , LAG(CAST(total_amount AS DOUBLE), 1, 0) OVER (PARTITION BY user_id ORDER BY order_date)  AS prev_amount
  , LEAD(CAST(total_amount AS DOUBLE), 1, 0) OVER (PARTITION BY user_id ORDER BY order_date) AS next_amount
  , total_amount - LAG(CAST(total_amount AS DOUBLE), 1, 0) OVER (PARTITION BY user_id ORDER BY order_date)
                                                                                             AS diff_amount
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,order_date,total_amount,prev_amount,next_amount,diff_amount
10001,2025-01-01,500.00,0.0,300.0,500.0
10001,2025-01-02,300.00,500.0,200.0,-200.0
10001,2025-01-03,200.00,300.0,800.0,-100.0
10001,2025-01-10,800.00,200.0,1200.0,600.0
10001,2025-01-15,1200.00,800.0,0.0,400.0
10002,2025-01-05,2000.00,0.0,1500.0,2000.0
10002,2025-01-06,1500.00,2000.0,3000.0,-500.0
10002,2025-01-07,3000.00,1500.0,5000.0,1500.0
10002,2025-01-20,5000.00,3000.0,0.0,2000.0
10007,2025-01-04,3000.00,0.0,2500.0,3000.0


### **6. 获取窗口首行 (FIRST_VALUE)**

<font color="red">题目：</font> 对于每个用户的每一笔订单，都标记出该用户的第一笔订单金额是多少。

<font color="#008000">考察点：</font>FIRST_VALUE()，无需修改窗口范围即可使用（默认从头开始）。

<font color="orange">答案：</font>

In [19]:
%%sql
SELECT
    user_id                                                                   AS user_id
  , order_date                                                                AS order_date
  , total_amount                                                              AS total_amount
  , FIRST_VALUE(total_amount) OVER (PARTITION BY user_id ORDER BY order_date) AS first_order_amt
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,order_date,total_amount,first_order_amt
10001,2025-01-01,500.00,500.00
10001,2025-01-02,300.00,500.00
10001,2025-01-03,200.00,500.00
10001,2025-01-10,800.00,500.00
10001,2025-01-15,1200.00,500.00
10002,2025-01-05,2000.00,2000.00
10002,2025-01-06,1500.00,2000.00
10002,2025-01-07,3000.00,2000.00
10002,2025-01-20,5000.00,2000.00
10007,2025-01-04,3000.00,3000.00


### **7. 获取窗口末行 (LAST_VALUE) - ⚠️陷阱题**

<font color="red">题目：</font> 对于每个用户的每一笔订单，都标记出该用户的最后一笔订单金额是多少。

<font color="#008000">考察点：</font>LAST_VALUE() 必须 配合 ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING，否则只能取到当前行。

<font color="orange">答案：</font>

In [20]:
%%sql
-- 注意：如果不加 ROWS 子句，last_order_amt 将永远等于 current total_amount
SELECT
    user_id                                                   AS user_id
  , order_date                                                AS order_date
  , total_amount                                              AS total_amount
  , LAST_VALUE(total_amount) OVER (PARTITION BY user_id ORDER BY order_date
    ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS last_order_amt
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,order_date,total_amount,last_order_amt
10001,2025-01-01,500.00,1200.00
10001,2025-01-02,300.00,1200.00
10001,2025-01-03,200.00,1200.00
10001,2025-01-10,800.00,1200.00
10001,2025-01-15,1200.00,1200.00
10002,2025-01-05,2000.00,5000.00
10002,2025-01-06,1500.00,5000.00
10002,2025-01-07,3000.00,5000.00
10002,2025-01-20,5000.00,5000.00
10007,2025-01-04,3000.00,5000.00


### **8. 获取指定行 (NTH_VALUE)**

<font color="red">题目：</font> 对于每个用户，找出其第二笔订单的金额，并填充到该用户的所有订单行中。

<font color="#008000">考察点：</font>NTH_VALUE(col, n)，同样需要注意窗口范围设置。

<font color="orange">答案：</font>

In [21]:
%%sql
SELECT
    user_id                                                   AS user_id
  , order_date                                                AS order_date
  , NTH_VALUE(total_amount, 2) OVER (PARTITION BY user_id ORDER BY order_date
    ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS second_order_amt
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,order_date,second_order_amt
10001,2025-01-01,300.00
10001,2025-01-02,300.00
10001,2025-01-03,300.00
10001,2025-01-10,300.00
10001,2025-01-15,300.00
10002,2025-01-05,1500.00
10002,2025-01-06,1500.00
10002,2025-01-07,1500.00
10002,2025-01-20,1500.00
10007,2025-01-04,2500.00


### **9. 累计求和 (SUM Window)**

<font color="red">题目：</font> 按日期顺序，计算全公司的订单金额累计总和（Running Total）。

<font color="#008000">考察点：</font>SUM() OVER (ORDER BY)，默认窗口即为从起点到当前行。

<font color="orange">答案：</font>

In [22]:
%%sql
SELECT
    TO_DATE(order_date)                                   AS dt
  , total_amount                                          AS total_amount
  , SUM(total_amount) OVER (ORDER BY TO_DATE(order_date)) AS running_total_gmv
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


dt,total_amount,running_total_gmv
2025-01-01,500.00,500.0
2025-01-02,300.00,1400.0
2025-01-02,600.00,1400.0
2025-01-03,200.00,2000.0
2025-01-03,400.00,2000.0
2025-01-04,3000.00,5000.0
2025-01-05,2000.00,9500.0
2025-01-05,2500.00,9500.0
2025-01-06,1500.00,15000.0
2025-01-06,4000.00,15000.0


### **10. 组内占比 (SUM Partition)**

<font color="red">题目：</font> 计算每个城市内，每个用户的消费金额占该城市总消费金额的比例。

<font color="#008000">考察点：</font>SUM() OVER (PARTITION BY)，分母为组内总和，分子为当前行。

<font color="orange">答案：</font>

In [23]:
%%sql
SELECT
    city                                                               AS city
  , user_id                                                            AS user_id
  , total_spent                                                        AS total_spent
  , SUM(total_spent) OVER (PARTITION BY city)                          AS city_total
  , ROUND(total_spent / SUM(total_spent) OVER ( PARTITION BY city), 4) AS contribution_rate
FROM
    (
    SELECT
        o.user_id           AS user_id
      , u.city              AS city
      , SUM(o.total_amount) AS total_spent
    FROM
        dataspire_catalog.db_dev.fact_orders o
            INNER JOIN dataspire_catalog.db_dev.dim_users u ON o.user_id = o.user_id
    GROUP BY
        1, 2
    ) t
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


city,user_id,total_spent,city_total,contribution_rate
深圳,10007,18000.0,41750.0,0.4311
深圳,10006,800.0,41750.0,0.0192
深圳,10008,1350.0,41750.0,0.0323
深圳,10004,1600.0,41750.0,0.0383
深圳,10001,3000.0,41750.0,0.0719
深圳,10005,4500.0,41750.0,0.1078
深圳,10002,11500.0,41750.0,0.2754
深圳,10003,1000.0,41750.0,0.024
上海,10002,23000.0,83500.0,0.2754
上海,10006,1600.0,83500.0,0.0192


### **11. 组内极值标记 (MAX/MIN Window)**

<font color="red">题目：</font> 找出每个用户的最大单笔订单金额，并判断当前订单是否为该用户的最大单（标记为 'Y' 或 'N'）。

<font color="#008000">考察点：</font>MAX() OVER (PARTITION BY)，结合 CASE WHEN 进行逻辑判断。

<font color="orange">答案：</font>

In [24]:
%%sql
SELECT
    user_id                                       AS user_id
  , order_id                                      AS order_id
  , total_amount                                  AS total_amount
  , MAX(total_amount) OVER (PARTITION BY user_id) AS max_order_amt
  , CASE WHEN total_amount = MAX(total_amount) OVER (PARTITION BY user_id) THEN 'Y'
         ELSE 'N'
    END                                           AS is_max_order
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,order_id,total_amount,max_order_amt,is_max_order
10001,202501010001,500.00,800.00,N
10001,202501020001,300.00,800.00,N
10001,202501030001,200.00,800.00,N
10001,202501100001,800.00,800.00,Y
10001,202501150001,1200.00,800.00,N
10002,202501050001,2000.00,5000.00,N
10002,202501060001,1500.00,5000.00,N
10002,202501070001,3000.00,5000.00,N
10002,202501200001,5000.00,5000.00,Y
10007,202501040001,3000.00,5000.00,N


## **五、 CTE 公用表表达式**

<font color="red">题目：</font> 先计算各渠道销售总额，再筛选出总额大于 2000 的渠道。

<font color="#008000">考察点：</font>WITH 子句 (CTE)，简化复杂查询逻辑，提高可读性。

<font color="orange">答案：</font>

In [25]:
%%sql
WITH channel_stats AS
         (
         SELECT
             channel                           AS channel
           , SUM(CAST(total_amount AS DOUBLE)) AS total_gmv
         FROM
             dataspire_catalog.db_dev.fact_orders
         WHERE
             status = 'paid'
         GROUP BY
             channel
         )
SELECT
    *
FROM
    channel_stats
WHERE
    total_gmv > 2000
ORDER BY
    total_gmv DESC
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


channel,total_gmv
app,31270.0
web,10080.0


## **六、 集合操作**
### **1. 全量合并保留重复 (UNION ALL)**

<font color="red">题目：</font> 将 2025-01-02 和 2025-01-03 两天的订单日志合并到一个结果集中。要求：即使两天内有完全相同的订单记录（如重试产生的重复日志），也要全部保留，不进行去重，以保证数据量的绝对真实。

<font color="#008000">考察点：</font>UNION ALL，性能最高（无 Shuffle 去重开销），适用于日志拼接、历史数据归档。

<font color="orange">答案：</font>

In [26]:
%%sql
-- 如果某订单在两天都出现（或同一天出现两次），结果中会显示多行。
SELECT
    order_id     AS order_id
  , user_id      AS user_id
  , total_amount AS total_amount
  , order_date   AS order_date
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-02'
UNION ALL
SELECT
    order_id     AS order_id
  , user_id      AS user_id
  , total_amount AS total_amount
  , order_date   AS order_date
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-03'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,user_id,total_amount,order_date
202501020001,10001,300.00,2025-01-02
202501020002,10003,600.00,2025-01-02
202501030001,10001,200.00,2025-01-03
202501030002,10003,400.00,2025-01-03


### **2. 合并并自动去重 (UNION)**

<font color="red">题目：</font> 将 2025-01-02 和 2025-01-03 两天的订单日志合并到一个结果集中。要求：即使两天内有完全相同的订单记录（如重试产生的重复日志），要进行去重操作。

<font color="#008000">考察点：</font>UNION (等同于 UNION DISTINCT)，会自动触发 Shuffle 和去重操作，性能略低于 UNION ALL。

<font color="orange">答案：</font>

In [27]:
%%sql
-- 模拟两个来源
SELECT
    order_id     AS order_id
  , user_id      AS user_id
  , total_amount AS total_amount
  , order_date   AS order_date
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-02'
UNION
SELECT
    order_id     AS order_id
  , user_id      AS user_id
  , total_amount AS total_amount
  , order_date   AS order_date
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-03'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,user_id,total_amount,order_date
202501020002,10003,600.00,2025-01-02
202501020001,10001,300.00,2025-01-02
202501030002,10003,400.00,2025-01-03
202501030001,10001,200.00,2025-01-03


### **3. 取交集 (INTERSECT)**

<font color="red">题目：</font> 找出在 2025-01-02 和2025-01-03都登录过的用户

<font color="#008000">考察点：</font>INTERSECT，返回两个查询结果集中共同存在的行（自动去重）。

<font color="orange">答案：</font>

In [28]:
%%sql
-- 结果：只返回那些在两个表中都出现的 user_id。
-- 等价写法：SELECT ... FROM A JOIN B ON A.id = B.id (但 INTERSECT 语义更清晰)
SELECT
    user_id AS user_id
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-02'
INTERSECT
SELECT
    user_id AS user_id
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-03'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id
10001
10003


### **4. 取差集 (EXCEPT / MINUS)**

<font color="red">题目：</font> 找出在 2025-01-02 下了单但没有注册过的“异常订单用户”（可能是代下单或系统漏洞）。

<font color="#008000">考察点：</font>EXCEPT (Spark 标准) 或 MINUS (Oracle/Hive 习惯，Spark 也兼容)，返回在左表中存在但在右表中不存在的行。

<font color="orange">答案：</font>

In [29]:
%%sql
-- 结果：返回有订单记录，但在登录表中找不到对应记录的用户。
-- 等价写法：LEFT ANTI JOIN (通常性能更好，但 EXCEPT 写法更简洁)
SELECT
    user_id AS user_id
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    order_date = '2025-01-02'
EXCEPT
SELECT
    user_id
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id


## **七、 透视表 (PIVOT) - Spark 3.0+**

<font color="red">题目：</font> 统计不同支付方式在各个月份的订单数量（行转列）。

<font color="#008000">考察点：</font>PIVOT 子句，行转列操作，聚合函数在透视中的应用。

<font color="orange">答案：</font>

In [30]:
%%sql
SELECT
    *
FROM
    (
    SELECT
        DATE_FORMAT(TO_DATE(order_date), 'yyyy-MM') AS month_str
      , payment_method
      , 1                                           AS cnt
    FROM
        dataspire_catalog.db_dev.fact_orders
    WHERE
        status = 'paid'
    ) t PIVOT (
        SUM(cnt) FOR payment_method IN ('alipay', 'wechat', 'credit', 'card')
    )
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


month_str,alipay,wechat,credit,card
2025-01,6,3,10,None
2025-02,4,2,None,None


## **八、 字符串函数**
### **1. 字符串拼接与格式化**

<font color="red">题目：</font> 查询用户表，将 user_name 和 city 拼接成格式为 “姓名-城市” 的字符串（例如：张三-北京），并重命名为 user_location。

<font color="#008000">考察点：</font>concat 函数，字符串字面量拼接。

<font color="orange">答案：</font>

In [31]:
%%sql
SELECT
    user_id                      AS user_id
  , CONCAT(user_name, '-', city) AS user_location
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_location
10001,张三-北京
10002,李四-上海
10003,王五-北京
10004,赵六-广州
10005,Emma-深圳
10006,John-上海
10007,Mike-北京
10008,Lisa-广州


### **2. 条件判断与标签化**

<font color="red">题目：</font> 查询用户表，根据 membership_level 给用户打上简单的中文标签：'钻石'->'至尊用户', '黄金'->'高级用户', 其他->'普通用户'

<font color="#008000">考察点：</font>CASE WHEN ... THEN ... ELSE ... END 语法，这是数据清洗中最常用的逻辑。

<font color="orange">答案：</font>

In [32]:
%%sql
SELECT
    user_id          AS user_id
  , user_name        AS user_name
  , membership_level AS membership_level
  , CASE WHEN membership_level = '钻石' THEN '至尊用户'
         WHEN membership_level = '黄金' THEN '高级用户'
         ELSE '普通用户'
    END              AS user_tag
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,user_name,membership_level,user_tag
10001,张三,黄金,高级用户
10002,李四,钻石,至尊用户
10003,王五,普通,普通用户
10004,赵六,白银,普通用户
10005,Emma,黄金,高级用户
10006,John,普通,普通用户
10007,Mike,钻石,至尊用户
10008,Lisa,白银,普通用户


### **3. 字符串截取 (提取年份)**

<font color="red">题目：</font> 从订单表的 order_id (如 202501010001) 中截取前 4 位作为“下单年份”，截取第 5-6 位作为“下单月份”。

<font color="#008000">考察点：</font>substr 或 substring 函数，理解索引位置（Spark SQL 通常从 1 开始）。

<font color="orange">答案：</font>

In [33]:
%%sql
SELECT
    order_id               AS order_id
  , SUBSTR(order_id, 1, 4) AS order_year
  , SUBSTR(order_id, 5, 2) AS order_month
FROM
    dataspire_catalog.db_dev.fact_orders
LIMIT 5
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,order_year,order_month
202501010001,2025,01
202501020001,2025,01
202501030001,2025,01
202501100001,2025,01
202501150001,2025,01


#### **4. 数值类型转换与计算**

<font color="red">题目：</font> 订单金额 total_amount 目前是 STRING 类型。请将其转换为 DOUBLE 类型，计算打 9 折后的价格，并保留 2 位小数。

<font color="#008000">考察点：</font>cast (类型转换), round (四舍五入), 算术运算。

<font color="orange">答案：</font>

In [34]:
%%sql
SELECT
    order_id                                     AS order_id
  , total_amount                                 AS original_amount
  , ROUND(CAST(total_amount AS DOUBLE) * 0.9, 2) AS discounted_price
FROM
    dataspire_catalog.db_dev.fact_orders
WHERE
    status = 'paid'
LIMIT 5
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,original_amount,discounted_price
202501010001,500.00,450.0
202501020001,300.00,270.0
202501030001,200.00,180.0
202501100001,800.00,720.0
202501150001,1200.00,1080.0


### **5. 空值填充 (COALESCE/NVL)**

<font color="red">题目：</font> 查询订单表，如果 discount_amount (优惠金额) 为 NULL，则显示为 0.00，否则显示原值。

<font color="#008000">考察点：</font>coalesce (返回第一个非空值) 或 nvl 函数，处理数据缺失。

<font color="orange">答案：</font>

In [35]:
%%sql
SELECT
    order_id                          AS order_id
  , discount_amount                   AS raw_discount
  , COALESCE(discount_amount, '0.00') AS safe_discount
FROM
    dataspire_catalog.db_dev.fact_orders
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_id,raw_discount,safe_discount
202501010001,50.00,50.00
202501020001,30.00,30.00
202501030001,20.00,20.00
202501100001,80.00,80.00
202501150001,100.00,100.00
202501050001,200.00,200.00
202501060001,150.00,150.00
202501070001,300.00,300.00
202501200001,500.00,500.00
202501020002,60.00,60.00


### **6. 多条件判断**

<font color="red">题目：</font> 根据 membership_level (会员等级) 生成一个新的列 level_score：'钻石' -> 4'黄金' -> 3'白银' -> 2'普通' -> 1其他 -> 0

<font color="#008000">考察点：</font>CASE WHEN ... THEN ... ELSE ... END 复杂逻辑判断。

<font color="orange">答案：</font>

In [36]:
%%sql
SELECT
    user_name        AS user_name
  , membership_level AS membership_level
  , CASE WHEN membership_level = '钻石' THEN 4
         WHEN membership_level = '黄金' THEN 3
         WHEN membership_level = '白银' THEN 2
         WHEN membership_level = '普通' THEN 1
         ELSE 0
    END              AS level_score
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_name,membership_level,level_score
张三,黄金,3
李四,钻石,4
王五,普通,1
赵六,白银,2
Emma,黄金,3
John,普通,1
Mike,钻石,4
Lisa,白银,2


## **九、 日期函数**
### **1. 日期解析与格式化**

<font color="red">题目：</font> 将订单表中的 order_date (STRING 类型，如 '2025-01-01') 转换为日期类型，并格式化为 "yyyy年MM月dd日" 的中文格式。

<font color="#008000">考察点：</font>to_date (转日期), date_format (格式化输出)。

<font color="orange">答案：</font>

In [37]:
%%sql
SELECT
    order_date                                         AS order_date
  , TO_DATE(order_date)                                AS date_obj
  , DATE_FORMAT(TO_DATE(order_date), 'yyyy年MM月dd日') AS cn_date
FROM
    dataspire_catalog.db_dev.fact_orders
LIMIT 5
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


order_date,date_obj,cn_date
2025-01-01,2025-01-01,2025年01月01日
2025-01-02,2025-01-02,2025年01月02日
2025-01-03,2025-01-03,2025年01月03日
2025-01-10,2025-01-10,2025年01月10日
2025-01-15,2025-01-15,2025年01月15日


### **2. 日期差值计算**

<font color="red">题目：</font> 计算每个用户的“注册天数”：即当前日期 (current_date()) 距离 register_date 的天数。

<font color="#008000">考察点：</font>current_date() (获取当天), datediff (计算日期差), to_date。

<font color="orange">答案：</font>

In [38]:
%%sql
SELECT
    user_name                                        AS user_name
  , register_date                                    AS register_date
  , DATEDIFF(CURRENT_DATE(), TO_DATE(register_date)) AS registered_days
FROM
    dataspire_catalog.db_dev.dim_users
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_name,register_date,registered_days
张三,2024-01-15,808
李四,2024-03-20,743
王五,2024-06-10,661
赵六,2025-01-05,452
Emma,2024-08-22,588
John,2024-11-30,488
Mike,2024-05-18,684
Lisa,2025-02-01,425


### **3. 字符串大小写转换与去空格**

<font color="red">题目：</font> 将 payment_method 转换为全大写，并假设数据中有空格，使用函数去除首尾空格（模拟清洗）

<font color="#008000">考察点：</font>upper/lower, trim。

<font color="orange">答案：</font>

In [39]:
%%sql
SELECT
    payment_method              AS raw_method
  , TRIM(UPPER(payment_method)) AS clean_method
FROM
    dataspire_catalog.db_dev.fact_orders
LIMIT 5
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


raw_method,clean_method
alipay,ALIPAY
wechat,WECHAT
alipay,ALIPAY
alipay,ALIPAY
credit,CREDIT


## **十、 数组与爆炸函数**

<font color="red">题目：</font> 模拟一个场景，将每个用户的订单ID聚合成一个数组，然后再次展开。

<font color="#008000">考察点：</font>collect_list (聚合为数组), explode (炸开数组), CTE 分步处理。

<font color="orange">答案：</font>

In [40]:
%%sql
-- 第一步：聚合为数组
WITH user_orders AS (
                    SELECT
                        user_id
                      , COLLECT_LIST(order_id) AS order_ids
                    FROM
                        dataspire_catalog.db_dev.fact_orders
                    GROUP BY
                        user_id
                    )
-- 第二步：炸开数组
SELECT
    user_id
  , EXPLODE(order_ids) AS single_order_id
FROM
    user_orders
WHERE
    user_id = '10001'
;

 * hive://64da22758f61d609:***@my-kyuubi-share-thrift-binary.spark.svc.cluster.local:10009/ecommerce?auth=LDAP
Done.


user_id,single_order_id
10001,202501010001
10001,202501020001
10001,202501030001
10001,202501100001
10001,202501150001
